File: oauth2/typescript/src/resilience/retry.ts
/**
 * OAuth2 Retry Executor
 *
 * Retry logic with exponential backoff for OAuth2 operations.
 */

import { OAuth2Error, NetworkError, ProviderError, isRetryable } from "../error";

/**
 * Retry configuration.
 */
export interface RetryConfig {
  /** Maximum number of retry attempts */
  maxRetries: number;
  /** Initial backoff delay in milliseconds */
  initialBackoffMs: number;
  /** Maximum backoff delay in milliseconds */
  maxBackoffMs: number;
  /** Backoff multiplier */
  backoffMultiplier: number;
  /** Jitter factor (0-1) */
  jitter: number;
}

/**
 * Default retry configuration.
 */
export const DEFAULT_RETRY_CONFIG: RetryConfig = {
  maxRetries: 3,
  initialBackoffMs: 500,
  maxBackoffMs: 30000,
  backoffMultiplier: 2.0,
  jitter: 0.1,
};

/**
 * Retry executor interface.
 */
export interface RetryExecutor {
  /**
   * Execute operation with retry.
   */
  execute<T>(operation: () => Promise<T>): Promise<T>;

  /**
   * Execute operation with retry and custom predicate.
   */
  executeIf<T>(
    operation: () => Promise<T>,
    shouldRetry: (error: Error, attempt: number) => boolean
  ): Promise<T>;
}

/**
 * OAuth2 retry executor implementation.
 */
export class OAuth2RetryExecutor implements RetryExecutor {
  private config: RetryConfig;

  constructor(config?: Partial<RetryConfig>) {
    this.config = { ...DEFAULT_RETRY_CONFIG, ...config };
  }

  async execute<T>(operation: () => Promise<T>): Promise<T> {
    return this.executeIf(operation, (error) => {
      if (error instanceof OAuth2Error) {
        return isRetryable(error);
      }
      return false;
    });
  }

  async executeIf<T>(
    operation: () => Promise<T>,
    shouldRetry: (error: Error, attempt: number) => boolean
  ): Promise<T> {
    let lastError: Error | undefined;
    let attempt = 0;

    while (attempt <= this.config.maxRetries) {
      try {
        return await operation();
      } catch (error) {
        lastError = error as Error;
        attempt++;

        // Check if we should retry
        if (attempt > this.config.maxRetries) {
          break;
        }

        if (!shouldRetry(lastError, attempt)) {
          break;
        }

        // Handle rate limit with Retry-After header
        const retryAfter = this.getRetryAfter(lastError);
        const backoff = retryAfter ?? this.calculateBackoff(attempt);

        await this.sleep(backoff);
      }
    }

    throw lastError;
  }

  private calculateBackoff(attempt: number): number {
    // Exponential backoff
    const baseDelay =
      this.config.initialBackoffMs *
      Math.pow(this.config.backoffMultiplier, attempt - 1);

    // Cap at max backoff
    const cappedDelay = Math.min(baseDelay, this.config.maxBackoffMs);

    // Add jitter
    const jitterAmount = cappedDelay * this.config.jitter;
    const jitter = (Math.random() * 2 - 1) * jitterAmount;

    return Math.round(cappedDelay + jitter);
  }

  private getRetryAfter(error: Error): number | undefined {
    if (error instanceof NetworkError && error.retryAfter) {
      return error.retryAfter * 1000; // Convert to milliseconds
    }
    return undefined;
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}

/**
 * Mock retry executor for testing.
 */
export class MockRetryExecutor implements RetryExecutor {
  private executeHistory: Array<{ attempts: number; success: boolean }> = [];
  private forceFailUntilAttempt: number = 0;

  /**
   * Force failure until specific attempt number.
   */
  setForceFailUntilAttempt(attempt: number): this {
    this.forceFailUntilAttempt = attempt;
    return this;
  }

  /**
   * Get execution history.
   */
  getExecuteHistory(): Array<{ attempts: number; success: boolean }> {
    return [...this.executeHistory];
  }

  async execute<T>(operation: () => Promise<T>): Promise<T> {
    return this.executeIf(operation, () => true);
  }

  async executeIf<T>(
    operation: () => Promise<T>,
    shouldRetry: (error: Error, attempt: number) => boolean
  ): Promise<T> {
    let attempts = 0;

    while (true) {
      attempts++;
      try {
        if (attempts <= this.forceFailUntilAttempt) {
          throw new NetworkError("Mock failure", "ConnectionFailed");
        }
        const result = await operation();
        this.executeHistory.push({ attempts, success: true });
        return result;
      } catch (error) {
        if (!shouldRetry(error as Error, attempts)) {
          this.executeHistory.push({ attempts, success: false });
          throw error;
        }
      }
    }
  }
}

/**
 * Create OAuth2 retry executor.
 */
export function createRetryExecutor(config?: Partial<RetryConfig>): RetryExecutor {
  return new OAuth2RetryExecutor(config);
}

/**
 * Create mock retry executor for testing.
 */
export function createMockRetryExecutor(): MockRetryExecutor {
  return new MockRetryExecutor();
}

File: oauth2/rust/src/resilience/retry.rs
//! Retry Logic
//!
//! Exponential backoff retry with jitter for OAuth2 operations.

use async_trait::async_trait;
use std::future::Future;
use std::sync::atomic::{AtomicU32, Ordering};
use std::sync::Mutex;
use std::time::Duration;

use crate::error::OAuth2Error;

/// Retry configuration.
#[derive(Debug, Clone)]
pub struct RetryConfig {
    /// Maximum number of retry attempts.
    pub max_attempts: u32,
    /// Initial delay before first retry.
    pub initial_delay: Duration,
    /// Maximum delay between retries.
    pub max_delay: Duration,
    /// Backoff multiplier.
    pub multiplier: f64,
    /// Jitter factor (0.0-1.0).
    pub jitter: f64,
}

impl Default for RetryConfig {
    fn default() -> Self {
        Self {
            max_attempts: 3,
            initial_delay: Duration::from_millis(100),
            max_delay: Duration::from_secs(10),
            multiplier: 2.0,
            jitter: 0.1,
        }
    }
}

/// Default retry configuration.
pub const DEFAULT_RETRY_CONFIG: RetryConfig = RetryConfig {
    max_attempts: 3,
    initial_delay: Duration::from_millis(100),
    max_delay: Duration::from_secs(10),
    multiplier: 2.0,
    jitter: 0.1,
};

/// Retry executor interface.
#[async_trait]
pub trait RetryExecutor: Send + Sync {
    /// Execute an operation with retry logic.
    async fn execute<T, F, Fut>(&self, operation: F) -> Result<T, OAuth2Error>
    where
        T: Send,
        F: Fn() -> Fut + Send + Sync,
        Fut: Future<Output = Result<T, OAuth2Error>> + Send;

    /// Check if an error is retryable.
    fn is_retryable(&self, error: &OAuth2Error) -> bool;

    /// Get retry statistics.
    fn get_stats(&self) -> RetryStats;
}

/// Retry statistics.
#[derive(Debug, Clone, Default)]
pub struct RetryStats {
    pub total_attempts: u32,
    pub successful_retries: u32,
    pub failed_operations: u32,
}

/// OAuth2 retry executor implementation.
pub struct OAuth2RetryExecutor {
    config: RetryConfig,
    stats: Mutex<RetryStats>,
}

impl OAuth2RetryExecutor {
    /// Create new retry executor.
    pub fn new(config: RetryConfig) -> Self {
        Self {
            config,
            stats: Mutex::new(RetryStats::default()),
        }
    }

    fn calculate_delay(&self, attempt: u32) -> Duration {
        let base_delay = self.config.initial_delay.as_millis() as f64
            * self.config.multiplier.powi(attempt as i32);

        let capped_delay = base_delay.min(self.config.max_delay.as_millis() as f64);

        // Add jitter
        let jitter_range = capped_delay * self.config.jitter;
        let jitter = (rand::random::<f64>() - 0.5) * 2.0 * jitter_range;
        let final_delay = (capped_delay + jitter).max(0.0);

        Duration::from_millis(final_delay as u64)
    }
}

impl Default for OAuth2RetryExecutor {
    fn default() -> Self {
        Self::new(RetryConfig::default())
    }
}

#[async_trait]
impl RetryExecutor for OAuth2RetryExecutor {
    async fn execute<T, F, Fut>(&self, operation: F) -> Result<T, OAuth2Error>
    where
        T: Send,
        F: Fn() -> Fut + Send + Sync,
        Fut: Future<Output = Result<T, OAuth2Error>> + Send,
    {
        let mut last_error = None;

        for attempt in 0..self.config.max_attempts {
            {
                let mut stats = self.stats.lock().unwrap();
                stats.total_attempts += 1;
            }

            match operation().await {
                Ok(result) => {
                    if attempt > 0 {
                        let mut stats = self.stats.lock().unwrap();
                        stats.successful_retries += 1;
                    }
                    return Ok(result);
                }
                Err(error) => {
                    if !self.is_retryable(&error) || attempt == self.config.max_attempts - 1 {
                        let mut stats = self.stats.lock().unwrap();
                        stats.failed_operations += 1;
                        return Err(error);
                    }

                    last_error = Some(error);
                    let delay = self.calculate_delay(attempt);
                    tokio::time::sleep(delay).await;
                }
            }
        }

        Err(last_error.unwrap_or_else(|| {
            OAuth2Error::Network(crate::error::NetworkError::Timeout {
                message: "Retry exhausted".to_string(),
            })
        }))
    }

    fn is_retryable(&self, error: &OAuth2Error) -> bool {
        error.is_retryable()
    }

    fn get_stats(&self) -> RetryStats {
        self.stats.lock().unwrap().clone()
    }
}

/// Mock retry executor for testing.
#[derive(Default)]
pub struct MockRetryExecutor {
    should_retry: Mutex<bool>,
    execution_count: AtomicU32,
    stats: Mutex<RetryStats>,
}

impl MockRetryExecutor {
    /// Create new mock executor.
    pub fn new() -> Self {
        Self::default()
    }

    /// Set whether to retry.
    pub fn set_should_retry(&self, should_retry: bool) -> &Self {
        *self.should_retry.lock().unwrap() = should_retry;
        self
    }

    /// Get execution count.
    pub fn get_execution_count(&self) -> u32 {
        self.execution_count.load(Ordering::SeqCst)
    }
}

#[async_trait]
impl RetryExecutor for MockRetryExecutor {
    async fn execute<T, F, Fut>(&self, operation: F) -> Result<T, OAuth2Error>
    where
        T: Send,
        F: Fn() -> Fut + Send + Sync,
        Fut: Future<Output = Result<T, OAuth2Error>> + Send,
    {
        self.execution_count.fetch_add(1, Ordering::SeqCst);
        operation().await
    }

    fn is_retryable(&self, _error: &OAuth2Error) -> bool {
        *self.should_retry.lock().unwrap()
    }

    fn get_stats(&self) -> RetryStats {
        self.stats.lock().unwrap().clone()
    }
}

/// Create retry executor.
pub fn create_retry_executor(config: RetryConfig) -> impl RetryExecutor {
    OAuth2RetryExecutor::new(config)
}

/// Create mock retry executor for testing.
pub fn create_mock_retry_executor() -> MockRetryExecutor {
    MockRetryExecutor::new()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_retry_config_default() {
        let config = RetryConfig::default();
        assert_eq!(config.max_attempts, 3);
        assert_eq!(config.initial_delay, Duration::from_millis(100));
    }

    #[tokio::test]
    async fn test_mock_executor() {
        let executor = MockRetryExecutor::new();

        let result = executor
            .execute(|| async { Ok::<_, OAuth2Error>("success") })
            .await;

        assert!(result.is_ok());
        assert_eq!(executor.get_execution_count(), 1);
    }
}
